blocks/api/raw-handling/utils.js

```diff
@@ -218,36 +128,133 @@ export function isPlain( HTML ) {
  * @param {NodeList} nodeList The nodeList to filter.
  * @param {Array}    filters  An array of functions that can mutate with the provided node.
  * @param {Document} doc      The document of the nodeList.
+ * @param {Object}   schema   The schema to use.
  */
-export function deepFilterNodeList( nodeList, filters, doc ) {
+export function deepFilterNodeList( nodeList, filters, doc, schema ) {
 	Array.from( nodeList ).forEach( ( node ) => {
-		deepFilterNodeList( node.childNodes, filters, doc );
+		deepFilterNodeList( node.childNodes, filters, doc, schema );
 
-		filters.forEach( ( filter ) => {
+		filters.forEach( ( item ) => {
 			// Make sure the node is still attached to the document.
 			if ( ! doc.contains( node ) ) {
 				return;
 			}
 
-			filter( node, doc );
+			item( node, doc, schema );
 		} );
 	} );
 }
 
 /**
  * Given node filters, deeply filters HTML tags.
+ * Filters from the deepest nodes to the top.
  *
  * @param {string} HTML    The HTML to filter.
  * @param {Array}  filters An array of functions that can mutate with the provided node.
+ * @param {Object} schema  The schema to use.
  *
  * @return {string} The filtered HTML.
  */
-export function deepFilterHTML( HTML, filters = [] ) {
+export function deepFilterHTML( HTML, filters = [], schema ) {
+	const doc = document.implementation.createHTMLDocument( '' );
+
+	doc.body.innerHTML = HTML;
+
+	deepFilterNodeList( doc.body.childNodes, filters, doc, schema );
+
+	return doc.body.innerHTML;
+}
+
+/**
+ * Given a schema, unwraps or removes nodes, attributes and classes on a node
+ * list.
+ *
+ * @param {NodeList} nodeList The nodeList to filter.
+ * @param {Document} doc      The document of the nodeList.
+ * @param {Object}   schema   An array of functions that can mutate with the provided node.
+ * @param {Object}   inline   Whether to clean for inline mode.
+ */
+function cleanNodeList( nodeList, doc, schema, inline ) {
+	Array.from( nodeList ).forEach( ( node ) => {
+		const tag = node.nodeName.toLowerCase();
+
+		// It's a valid child.
+		if ( schema.hasOwnProperty( tag ) || schema.hasOwnProperty( node.nodeType ) ) {
+			if ( node.nodeType === ELEMENT_NODE ) {
+				const { attributes = [], classes = [], children, require = [] } = schema[ tag ];
+
+				// If the node is empty and it's supposed to have children,
+				// remove the node.
+				if ( isEmpty( node ) && children ) {
+					remove( node );
+					return;
+				}
+
+				// Strip invalid attributes.
+				Array.from( node.attributes ).forEach( ( { name } ) => {
+					if ( name !== 'class' && attributes.indexOf( name ) === -1 ) {
+						node.removeAttribute( name );
+					}
+				} );
+
+				// Strip invalid classes.
+				const newClasses = classes.filter( ( name ) => node.classList.contains( name ) );
```

Comment:
Should we check to see whether there are even any classes assigned before running through this filter / removing the attribute?

https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/length